MBT Articulated Body Algorithm

bobbyluig opened the pull request:
This PR completes the implementation of the articulated body algorithm for computing forward dynamics in MBT. It also includes tests using the Kuka iiwa and double pendulum.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8194)
<!-- Reviewable:end -->

amcastro-tri commented:
+@amcastro-tri for feature review.

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-:-L6OHyOIDYbYY_cNaLrI:b-dsacsr)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
This is great! Quick checkpoint.

---

Reviewed 2 of 12 files at r1.
Review status: 2 of 12 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 253 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YErhvrfIHK6gepjM-:-L6YErhvrfIHK6gepjM0:bwexks8) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L253)):*
> ```objc
> 
>     // Compute forward dynamics using articulated body algorithm.
>     model_->CalcForwardDynamics(*context_, pc, aic, aac, qDDt);
> ```

I'd make the signature for this method so I only need to call: "model_->CalcForwardDynamics(const Context&, const MultibodyForces&, EigenPtr<VectorX>)", i.e. place all the cache evaluation (meaning EvalXXX() calls) within that method, outside users' reach. This will help a lot to stabilize the API.

---

*[multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_forward_dynamics_test.cc, line 9 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6Y6FlMD0HpywjRFBCf:-L6Y6FlMD0HpywjRFBCg:b-x3uz7q) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_forward_dynamics_test.cc#L9)):*
> ```cpp
> namespace multibody {
> namespace benchmarks {
> namespace kuka_iiwa_robot {
> ```

This tests are awesome @bobbyluig!
It'd be nice to have them somewhere in `drake/multibody/multibody_tree/test/multibody_tree_forward_dynamics_test.cc` instead since they test MultibodyTree, not the benchmarks. Also, taking `multibody_tree_test.cc` as an example, where I moved away from using the class `DrakeKukaIIwaRobot` and started using the factory `MakeKukaIiwaModel()`. I like that approach in that the tests themselves look clean enough that they also serve as pedagogical examples we can use to point people into so they can learn how to use these methods. Going through `DrakeKukaIIwaRobot` (we did that in the past as a quick hack to get MBT tests in but I might remove it soon) requires people to go through that extra layer to see the actual MBT calls. 

---

*[multibody/multibody_tree/articulated_body_inertia_cache.cc, line 5 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6Y8h3GB7wdtUi1Eg-6:-L6Y8h3GB7wdtUi1Eg-7:bmugehw) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/articulated_body_inertia_cache.cc#L5)):*
> ```cpp
> #include "drake/common/default_scalars.h"
> 
> DRAKE_DEFINE_CLASS_TEMPLATE_INSTANTIATIONS_ON_DEFAULT_SCALARS(
> ```

BTW nice!

---

*[multibody/multibody_tree/articulated_body_inertia_cache.h, line 15 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YBbVpKoirvwkXnAnQ:-L6YBbVpKoirvwkXnAnR:bwoz0os) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/articulated_body_inertia_cache.h#L15)):*
> ```objc
> /// This class is one of the cache entries in MultibodyTreeContext. It holds the
> /// results of computations that are used in the recursive implementation of the
> /// articulated body algorithm.
> ```

I would update this doc now that we stabilize on what to place here. Something like:
"This class is one of the cache entries in MultibodyTreeContext. It is used to store the results from the first pass of the articulated body algorithm for the computation of articulated body inertias. In addition to storing the articulated body inertias themselves, this cache entry also serves to store quantities computed during the first pass but that are also needed during the second pass. Quantities in this cache entry are function of the generalized position vector q only."

---

*[multibody/multibody_tree/articulated_body_inertia_cache.h, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6Y8RP9QnsGqOskPc5W:-L6Y8RP9QnsGqOskPc5X:bttm00x) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/articulated_body_inertia_cache.h#L18)):*
> ```objc
> ///
> /// Articulated body inertia cache entries include:
> /// - Articulated body inertia `P_B_W` of the body taken about Bo and expressed
> ```

BTW "of the body " --> " of body B "

---

*[multibody/multibody_tree/articulated_body_inertia_cache.h, line 123 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YB0BHw2LColPnkZNb:-L6YB0BHw2LColPnkZNc:b-iy0pa0) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/articulated_body_inertia_cache.h#L123)):*
> ```objc
>     Pplus_PB_W_.resize(num_nodes_);
>     ldlt_D_B.resize(num_nodes_);
>     g_PB_W.resize(num_nodes_);
> ```

I think a good idea would be to initialize these at `[world_index()]` to NaN values so that accessing and using them does not go unnoticed in case they are set to zero. We do that in the other kinematic cache structures as well (more than once I got NaN for mistakenly using these values which helped finding bugs).

---

*[multibody/multibody_tree/body_node.h, line 1016 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6Y9zqI8sJUVzncKBvx:-L6Y9zqI8sJUVzncKBvy:bkeugcq) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/body_node.h#L1016)):*
> ```objc
>     // Compute the LDLT factorization of D_B as ldlt_D_B.
>     // TODO(bobbyluig): Test performance against inverse().
>     auto& ldlt_D_B = get_mutable_ldlt_D_B(aic);
> ```

`Eigen::LDLT<MatrixUpTo6<T>>&`?, We prefer to spell things out when the return type is not obvious. (E.g, it is obvious in `auto p = std::make_unique<Foo>()`)

---

*[multibody/multibody_tree/multibody_tree.cc, line 859 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6Y9PXCyYvqmkvGDfRs:-L6Y9PXCyYvqmkvGDfRt:b-g8lxro) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.cc#L859)):*
> ```cpp
>     const systems::Context<T>& context,
>     const PositionKinematicsCache<T>& pc,
>     ArticulatedBodyInertiaCache<T>* aic) const {
> ```

I liked `abc` :-(. Maybe `abic`?

---

*[multibody/multibody_tree/multibody_tree.cc, line 888 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6Y8zOLw1kt8EFD7Fq6:-L6Y8zOLw1kt8EFD7Fq7:b9cisci) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.cc#L888)):*
> ```cpp
>     const PositionKinematicsCache<T>& pc,
>     const VelocityKinematicsCache<T>& vc,
>     const ArticulatedBodyInertiaCache<T>& aic,
> ```

BTW maybe better `abic`?

---

*[multibody/multibody_tree/multibody_tree.h, line 1423 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6Y8--t0uNFQu-5KzX_:-L6Y8--t0uNFQu-5KzXa:bwpm16n) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.h#L1423)):*
> ```objc
>   /// Computes some quantities that are required in the final pass of the
>   /// articulated body algorithm which depend only on generalized positions and
>   /// stores them in the articulated body inertia cache `aic`.
> ```

I would make this method `private:`, users do not need to call it themselves but `MBT::CalcForwardDynamics()` does. 
The doc should probably say that this method computes articulated body inertia and in addition it caches a number of quantities used in the next passes of the articulated body algorithm.

---

*[multibody/multibody_tree/multibody_tree.h, line 1445 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YG2aJVC3B3n-h1u6P:-L6YG2aK5YfdErcTyB1O:b-qa8afd) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.h#L1445)):*
> ```objc
>   /// @pre The position kinematics `pc` must have been previously updated with a
>   /// call to CalcPositionKinematicsCache() using the same `context`.
>   void CalcArticulatedBodyInertiaCache(
> ```

We are in the position of making this `private:` now. Users only care about calling `CalcForwardDynamics()`, caching evaluation happens under the hood automatically for them.

---

*[multibody/multibody_tree/multibody_tree.h, line 1491 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YGJSOEdaF90LDJvHc:-L6YGJSOEdaF90LDJvHd:b-qa8afd) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.h#L1491)):*
> ```objc
>   /// @pre The multibody forces `forces` must have been previously updated with
>   /// a call to CalcForceElementsContribution() using the same `context`.
>   void CalcArticulatedBodyAlgorithmCache(
> ```

We are in the position of making this `private:` now. Users only care about calling `CalcForwardDynamics()`, caching evaluation happens under the hood automatically for them.

---

*[multibody/multibody_tree/multibody_tree.h, line 1501 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YGnK84-xwJK4WpCSC:-L6YGnK84-xwJK4WpCSD:b-tkzimm) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.h#L1501)):*
> ```objc
>   /// Compute the generalized accelerations for this %MultibodyTree model using
>   /// the articulated body algorithm.
>   ///
> ```

I would be more generous with the doc here. A number of beautiful qualitities are missing, like this method is only `O(n)`, you really need to brag about that! Also let's acknowledge Featherstone for this beauty. 

Something like this (please cleanup my broken English):
"Given a state for `this` model stored in `context` and a set of applied forces (`applied_forces`), this method uses the`O(n)` articulated body algorithm developed by [Featherstone 1983] to compute generalized accelerations `vdot`, that is the time derivatives of the generalized velocities v stored in `context`.

All the params docs....

References: (maybe others if you feel there are other important ones that helped you out).
- [Featherstone 1983] Featherstone, R., 1983.
    The calculation of robot dynamics using articulated-body inertias. The
    International Journal of Robotics Research, 2(1), pp.13-30.

"

---

*[multibody/multibody_tree/multibody_tree.h, line 1510 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YLh7J3WAUM95oQepA:-L6YLh7J3WAUM95oQepB:b-24oaxj) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.h#L1510)):*
> ```objc
>   ///   An articulated body inertia cache object already updated to be in sync
>   ///   with `context`.
>   /// @param[in] aac
> ```

BTW lets get rid of these cache arguments so that it's clean to users.

---

*[multibody/multibody_tree/multibody_tree.h, line 1530 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YFVyt16lJbH8ullDV:-L6YFVyt16lJbH8ullDW:b-wu5oxs) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.h#L1530)):*
> ```objc
>       const ArticulatedBodyInertiaCache<T>& aic,
>       const ArticulatedBodyAlgorithmCache<T>& aac,
>       EigenPtr<VectorX<T>> vdot) const;
> ```

I'd simplify this API to:
```c++
void  CalcForwardDynamics(
      const systems::Context<T>& context,
      const MultibodyForces<T>& applied_forces,  // Only external forces here, internally we add force elements.
      EigenPtr<VectorX<T>> vdot) const;
```

and bury all the cache evaluations inside this method. (There are EvalPositionKinematics() and EvalVelocityKinematics() for you to use in there now).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-:-L6Ycywv-kpYHQ9pVyih:bl27p34)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 2 of 12 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[multibody/multibody_tree/multibody_tree.h, line 1501 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YGnK84-xwJK4WpCSC:-L6YojZT81OxgC-pRWaj:bmxj8yx) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.h#L1501)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I would be more generous with the doc here. A number of beautiful qualitities are missing, like this method is only `O(n)`, you really need to brag about that! Also let's acknowledge Featherstone for this beauty. 

Something like this (please cleanup my broken English):
"Given a state for `this` model stored in `context` and a set of applied forces (`applied_forces`), this method uses the`O(n)` articulated body algorithm developed by [Featherstone 1983] to compute generalized accelerations `vdot`, that is the time derivatives of the generalized velocities v stored in `context`.

All the params docs....

References: (maybe others if you feel there are other important ones that helped you out).
- [Featherstone 1983] Featherstone, R., 1983.
    The calculation of robot dynamics using articulated-body inertias. The
    International Journal of Robotics Research, 2(1), pp.13-30.

"

</blockquote></details>

BTW although I _think_ Featherstone came up with the name "Articulated Body Inertia" several people invented the concept separately. Here I think we are using the algorithm and terminology from Abhi Jain, which was developed independently starting from a completely different idea, Kalman filtering. So Jain should at least get _some_ credit here!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8194)*
<!-- Sent from Reviewable.io -->

bobbyluig commented:
Review status: 1 of 11 files reviewed at latest revision, 15 unresolved discussions.

---

*[multibody/multibody_tree/articulated_body_inertia_cache.h, line 15 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YBbVpKoirvwkXnAnQ:-L6iAgck_K2At6X5yUsB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/articulated_body_inertia_cache.h#L15)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I would update this doc now that we stabilize on what to place here. Something like:
"This class is one of the cache entries in MultibodyTreeContext. It is used to store the results from the first pass of the articulated body algorithm for the computation of articulated body inertias. In addition to storing the articulated body inertias themselves, this cache entry also serves to store quantities computed during the first pass but that are also needed during the second pass. Quantities in this cache entry are function of the generalized position vector q only."
</blockquote></details>

Done.

---

*[multibody/multibody_tree/articulated_body_inertia_cache.h, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6Y8RP9QnsGqOskPc5W:-L6iAWaHzNB_QJbn3acZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/articulated_body_inertia_cache.h#L18)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW "of the body " --> " of body B "
</blockquote></details>

Done.

---

*[multibody/multibody_tree/articulated_body_inertia_cache.h, line 123 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YB0BHw2LColPnkZNb:-L6he2_qEOyagD4xB_fr:b1dxi5e) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/articulated_body_inertia_cache.h#L123)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I think a good idea would be to initialize these at `[world_index()]` to NaN values so that accessing and using them does not go unnoticed in case they are set to zero. We do that in the other kinematic cache structures as well (more than once I got NaN for mistakenly using these values which helped finding bugs).
</blockquote></details>

I'm running into the issue where symbolic doesn't work well with LDLT. Any way around this?

---

*[multibody/multibody_tree/body_node.h, line 1016 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6Y9zqI8sJUVzncKBvx:-L6hem0qmwAUFbpXaJ3G:bs7wumb) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/body_node.h#L1016)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

`Eigen::LDLT<MatrixUpTo6<T>>&`?, We prefer to spell things out when the return type is not obvious. (E.g, it is obvious in `auto p = std::make_unique<Foo>()`)
</blockquote></details>

Done.

---

*[multibody/multibody_tree/multibody_tree.cc, line 859 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6Y9PXCyYvqmkvGDfRs:-L6hkfcbG5WrS3dzMjlB:b5qfs39) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.cc#L859)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I liked `abc` :-(. Maybe `abic`?
</blockquote></details>

I changed this because `abc` doesn't hint at the difference between `InertiaCache` and `AlgorithmCache. Using `abic`.

---

*[multibody/multibody_tree/multibody_tree.cc, line 888 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6Y8zOLw1kt8EFD7Fq6:-L6hllflU1u3Qcm-9dIz:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.cc#L888)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW maybe better `abic`?
</blockquote></details>

Done.

---

*[multibody/multibody_tree/multibody_tree.h, line 1423 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6Y8--t0uNFQu-5KzX_:-L6iAA649pb01jgrLO9K:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.h#L1423)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I would make this method `private:`, users do not need to call it themselves but `MBT::CalcForwardDynamics()` does. 
The doc should probably say that this method computes articulated body inertia and in addition it caches a number of quantities used in the next passes of the articulated body algorithm.
</blockquote></details>

Done.

---

*[multibody/multibody_tree/multibody_tree.h, line 1445 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YG2aJVC3B3n-h1u6P:-L6hlptr5aXrAKw7eCxh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.h#L1445)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

We are in the position of making this `private:` now. Users only care about calling `CalcForwardDynamics()`, caching evaluation happens under the hood automatically for them.
</blockquote></details>

Done.

---

*[multibody/multibody_tree/multibody_tree.h, line 1491 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YGJSOEdaF90LDJvHc:-L6hlqAfuUdDpv5hNZdA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.h#L1491)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

We are in the position of making this `private:` now. Users only care about calling `CalcForwardDynamics()`, caching evaluation happens under the hood automatically for them.
</blockquote></details>

Done.

---

*[multibody/multibody_tree/multibody_tree.h, line 1501 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YGnK84-xwJK4WpCSC:-L6iW35oFD3iC657je3S:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.h#L1501)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW although I _think_ Featherstone came up with the name "Articulated Body Inertia" several people invented the concept separately. Here I think we are using the algorithm and terminology from Abhi Jain, which was developed independently starting from a completely different idea, Kalman filtering. So Jain should at least get _some_ credit here!
</blockquote></details>

Done.

---

*[multibody/multibody_tree/multibody_tree.h, line 1510 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YLh7J3WAUM95oQepA:-L6hinaoZEJYX-ogEnw3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.h#L1510)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW lets get rid of these cache arguments so that it's clean to users.
</blockquote></details>

Done.

---

*[multibody/multibody_tree/multibody_tree.h, line 1530 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YFVyt16lJbH8ullDV:-L6hfl5jyb0hG9QCLL0a:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/multibody_tree.h#L1530)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I'd simplify this API to:
```c++
void  CalcForwardDynamics(
      const systems::Context<T>& context,
      const MultibodyForces<T>& applied_forces,  // Only external forces here, internally we add force elements.
      EigenPtr<VectorX<T>> vdot) const;
```

and bury all the cache evaluations inside this method. (There are EvalPositionKinematics() and EvalVelocityKinematics() for you to use in there now).
</blockquote></details>

Done.

---

*[multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 253 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YErhvrfIHK6gepjM-:-L6iLmLOKiEezFye-ogR:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L253)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I'd make the signature for this method so I only need to call: "model_->CalcForwardDynamics(const Context&, const MultibodyForces&, EigenPtr<VectorX>)", i.e. place all the cache evaluation (meaning EvalXXX() calls) within that method, outside users' reach. This will help a lot to stabilize the API.
</blockquote></details>

Done.

---

*[multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_forward_dynamics_test.cc, line 9 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6Y6FlMD0HpywjRFBCf:-L6iPHjn_gvvDexKhGhh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_forward_dynamics_test.cc#L9)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This tests are awesome @bobbyluig!
It'd be nice to have them somewhere in `drake/multibody/multibody_tree/test/multibody_tree_forward_dynamics_test.cc` instead since they test MultibodyTree, not the benchmarks. Also, taking `multibody_tree_test.cc` as an example, where I moved away from using the class `DrakeKukaIIwaRobot` and started using the factory `MakeKukaIiwaModel()`. I like that approach in that the tests themselves look clean enough that they also serve as pedagogical examples we can use to point people into so they can learn how to use these methods. Going through `DrakeKukaIIwaRobot` (we did that in the past as a quick hack to get MBT tests in but I might remove it soon) requires people to go through that extra layer to see the actual MBT calls. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8194)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
This is great work @bobbyluig! I'm still working my way through it.

---

Reviewed 1 of 12 files at r1, 7 of 12 files at r2.
Review status: 7 of 11 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[multibody/multibody_tree/articulated_body_algorithm_cache.h, line 88 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6sQ3EYqVbev2BIUYbU:-L6sQ3EYqVbev2BIUYbV:boy4bb4) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/articulated_body_algorithm_cache.h#L88)):*
> ```objc
>     return e_B_[body_node_index];
>   }
> ```

BTW blank line

---

*[multibody/multibody_tree/articulated_body_algorithm_cache.h, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6sSHrdhBh6dF1USOzp:-L6sSHrdhBh6dF1USOzq:b2w54d6) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/articulated_body_algorithm_cache.h#L111)):*
> ```objc
> 
>   // Pools.
>   SpatialForce_PoolType zplus_PB_W_{};  // Indexed by BodyNodeIndex.
> ```

BTW maybe place this comment in the line above just once instead of repeating on each line.

---

*[multibody/multibody_tree/articulated_body_inertia_cache.h, line 123 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YB0BHw2LColPnkZNb:-L6qiQrx3gPbwGUv4uOW:b-pca82w) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/articulated_body_inertia_cache.h#L123)):*
<details><summary><i>Previously, bobbyluig (Lujing Cen) wrote…</i></summary><blockquote>

I'm running into the issue where symbolic doesn't work well with LDLT. Any way around this?
</blockquote></details>

because of this initialization to NaN? where specifically? I could take a look at it.

---

*[multibody/multibody_tree/body_node.h, line 1048 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6sRTOg5ZNgRfzVE8_l:-L6sRTOg5ZNgRfzVE8_m:b-gkaq97) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1048)):*
> ```objc
> 
>   /// This method is used by MultibodyTree within a tip-to-base loop to compute
>   /// this node's articulated body algorithm quantities that depend only both
> ```

BTW only --> on

---

*[multibody/multibody_tree/body_node.h, line 1115 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6sXkFbg7owKlSmxETb:-L6sXkFbg7owKlSmxETc:bp57m2d) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1115)):*
> ```objc
>     //  - z_Bo_W for the residual spatial force, such that we can write
>     //    F_Bo_W = P_B_W A_WB + z_Bo_W.
>     //  - zplus_PB_W for the articulated body inertia residual force but
> ```

It doesn't seem like we need an extra symbol here. Recall that Jain uses `+` to denote the frame immediately inboard of the current body frame B. I think that frame for us is `Pb`, that is, body frame P shifted to be at Bo. Therefore I think instead of zplus (which is a spatial force) we could just use z_PPb_W (spatial force on P, at Po, expressed in W). That is, there's nothing special about zplus, it's just another bias term but at another frame (I think?). Look at Eq. 6.35 in page 108 of Jain's book, which says `z(k+1) = phi(k+1,k)*zplus(k)`. With this notation we'd simply get `Z_PPo_W = Z_PPb_W.Shift(p_BoPo_W)`.
Also, and this is just preference, maybe since z is a spatial force we could use upper case Z as we do for the upper case symbol F, i.e. `Z_PPb_W`. 

---

*[multibody/multibody_tree/body_node.h, line 1332 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6s-tYPDqiJMmqH3G8J:-L6s-tYPDqiJMmqH3G8K:b-ajfkak) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1332)):*
> ```objc
>     //  - B body frame associated with this node.
>     //  - P ("parent") body frame associated with this node's parent.
>     //  - Aplus_WP for the rigidly propagated spatial acceleration from the
> ```

Jain calls this Aplus because he counts from 1 to n from tip to base. This acceleration however is the acceleration of frame Pb (frame P shifted to b) if the system was at rest. Maybe we could call it something like `A0_WPb`?, that is, the acceleration of frame P shifted to B when the system is at rest (and thus velocities are zero).

---

*[multibody/multibody_tree/body_node.h, line 1345 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6qpZYW70cp6ECLXsNn:-L6qpZYXMF2WGORKdYBx:b674c8r) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1345)):*
> ```objc
>     // We first rigidly propagate the spatial acceleration of the parent to the
>     // body frame. Note that the world frame is assumed to have already been
>     // initialized with zero spatial acceleration. We define p_PoBo_W to be the
> ```

BTW, is a "zero acceleration" a pre-requisite? I mean, could this entire algorithm work in a non-inertial frame of reference where the acceleration of the world is non-zero? For instance, imagine your robot is operating inside a moving vehicle (or the robot IS the vehicle). 

---

*[multibody/multibody_tree/body_node.h, line 1351 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6qrlfPBcZTLiup_c6X:-L6qrlfPBcZTLiup_c6Y:b-bdn9ku) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1351)):*
> ```objc
>     //
>     // Using Aplus_WP, we can compute the generalized acceleration for this
>     // body's mobilizer. Note that we assume that any gravity has already been
> ```

is there a reason to think about gravity differently from other forces? maybe say: "Note that we assume that any external forces have already been incorporated in the previous pass of the articulated body algorithm."

---

*[multibody/multibody_tree/multibody_tree.cc, line 755 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6qmQuWFyXnH92FchTg:-L6qmQuWFyXnH92FchTh:boy4bb4) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/multibody_tree.cc#L755)):*
> ```cpp
>     // Position of origin Fo measured from Bi, expressed in the world W.
>     //
>     const Vector3<T> p_BiFo_W = p_WoFo_W - p_WBi;
> ```

BTW blank line

---

*[multibody/multibody_tree/multibody_tree.cc, line 969 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6qmyMcyyI3xIU1pRBB:-L6qmyMcyyI3xIU1pRBC:b1s2nik) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/multibody_tree.cc#L969)):*
> ```cpp
> 
>       node.CalcForwardDynamics_BaseToTip(
>           mbt_context, pc, abic, abac, H_PB_W, &ac, vdot);
> ```

Actually, this is the perfect place for the acceleration cache (I initially had messed up how to use it).
So, on a second thought, why not to place `vdot` within the acceleration kinematics cache, and make the signature of the forward dynamics `CalcForwardDynamics(const Context&, const MultibodyForces&, AccelerationKinematicsCache*)`? That also gives us a chance to not only save `vdot` by also `A_WB`

---

*[multibody/multibody_tree/multibody_tree.h, line 1445 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6qj7ojGYH_dh9-Zsvc:-L6qj7ojGYH_dh9-Zsvd:b-aedts) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/multibody_tree.h#L1445)):*
> ```objc
>   /// - [Jain 2010] Jain, A., 2010. Robot and multibody dynamics: analysis and
>   ///               algorithms. Springer Science & Business Media, pp. 123-130.
>   /// - [Featherstone 2008] Featherstone, R., 2008. Rigid body dynamics
> ```

BTW, not cited in the text. Did you mean to?

---

*[multibody/multibody_tree/multibody_tree.h, line 1891 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6sQbSB9eXrz7qspEG8:-L6sQbSB9eXrz7qspEG9:b-745v7f) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/multibody_tree.h#L1891)):*
> ```objc
>   // Computes some quantities that are required in the final pass of the
>   // articulated body algorithm which depend only on generalized positions and
>   // stores them in the articulated body inertia cache `abic`.
> ```

This and the doc for `CalcArticulatedBodyAlgorithmCache()` are very similar. I think it is worth mentioning in these docs that `CalcArticulatedBodyInertiaCache()` only depends on q and therefore solving forward dynamics with different forces, but the same context,  would require multiple evaluations of `CalcArticulatedBodyAlgorithmCache()` but only a single call to `CalcArticulatedBodyInertiaCache()`

---

*[multibody/multibody_tree/test/tree_from_joints_test.cc, line 381 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6qkDJ1_vEvA9lxgupF:-L6qkDJ1_vEvA9lxgupG:b9q269j) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/test/tree_from_joints_test.cc#L381)):*
> ```cpp
>     Vector2d qddot_expected;
> 
>     // Test with no angular velocity.
> ```

BTW comment is not correct?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-:-L6sZEkEi9OkURZ5b-5j:bqu79gw)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 7 of 11 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[multibody/multibody_tree/body_node.h, line 1133 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6vd42lDd9qJWJW4kYn:-L6vd42lDd9qJWJW4kYo:b-pamilb) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1133)):*
> ```objc
>     //   z_Bo_W = Σᵢ(Φ(p_BCᵢ_W) zplus_BCᵢ_W)
>     //            + P_B_W a_Bo_W + b_Bo_W - Fapplied_Bo_W
>     //          = Σᵢ(zplus_BCᵢb_W) + P_B_W a_Bo_W + b_Bo_W - Fapplied_Bo_W  (1)
> ```

The derivations for these equations are quite lengthy but, could you at least cite an Eq. number in Jain's book?

---

*[multibody/multibody_tree/body_node.h, line 1186 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6vgnD0FYB3Hn6hsxlQ:-L6vgnD0FYB3Hn6hsxlR:b-h4ajkq) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1186)):*
> ```objc
> 
>     // Compute the gyroscopic spatial force using (2).
>     // TODO(bobbyluig): Get value from cache when available.
> ```

BTW could you add to TODO comment this also gets used in `BodyNode::CalcBodySpatialForceGivenItsSpatialAcceleration()`?

---

*[multibody/multibody_tree/body_node.h, line 1187 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6vgL41nRgdpSkzBqkA:-L6vgL41nRgdpSkzBqkB:bbedj99) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1187)):*
> ```objc
>     // Compute the gyroscopic spatial force using (2).
>     // TODO(bobbyluig): Get value from cache when available.
>     SpatialForce<T> b_Bo_W = mass *
> ```

BTW `const`

---

*[multibody/multibody_tree/body_node.h, line 1218 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6viFEOvqUeFUsSP_wA:-L6viFEOvqUeFUsSP_wB:bbhlsi9) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1218)):*
> ```objc
> 
>     // Get Azero_PB_W by shifting and re-expressing Azero_FM.
>     const SpatialAcceleration<T> Azero_PB_W =
> ```

BTW nice! very well done. It just needs some brief doc for some future developer going through this.

---

*[multibody/multibody_tree/body_node.h, line 1219 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6vhhPANckfrrytuS13:-L6vhhPANckfrrytuS14:b-sz3kd7) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1219)):*
> ```objc
>     // Get Azero_PB_W by shifting and re-expressing Azero_FM.
>     const SpatialAcceleration<T> Azero_PB_W =
>         R_WF * Azero_FM.Shift(p_MoBo_F, V_FM.rotational());
> ```

This is not immediately obvious to someone who has not done this before. Could you add brief comment on how/why using `w_FM` instead of `w_PB`? (are they equal? why? assumptions).
Also comment on how the F frame in `Azero_FM` becomes the P frame in `Azero_PB_W` (that is, why is it that `A_PM = A_FM`?) (I actually had seconds thoughts myself for a moment).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8194)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 7 of 11 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[multibody/multibody_tree/body_node.h, line 1235 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L75cE7SemZ_j-NtlTej:-L75cE7SemZ_j-NtlTek:b-fcovkr) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1235)):*
> ```objc
>     a_Bo_W = SpatialAcceleration<T>(
>         w_WB.cross(w_PB_W) + Azero_PB_W.rotational(),
>         w_WP.cross(v_WB - v_WP + v_PB_W) + Azero_PB_W.translational());
> ```

Ok, I am pretty certain (I will push derivations soon to add here in this PR), that the Coriolis term here is off by a factor of 2. Regarding the rotational term, notice that: `w_WB x w_PB_W = w_WP x w_PB_W` and therefore that terms is nothing but the cross velocity term appearing when composing accelerations (more on this later when I update the notes).
Regarding the translational term notice that `v_WB = v_WP + w_WP x p_PoBo_W` and therefore the rotational component is: `w_WP x (v_WB - v_WP + v_PB_W) = w_WP x w_WP x p_PoBo_W + w_WP x v_PB_W`. The first part on the right hand side is the centrifugal acceleration while the second part should correspond to the Coriolis acceleration. However, Coriolis always comes with a factor of 2, i.e I believe this term should be `w_WP x w_WP x p_PoBo_W + 2 * w_WP x v_PB_W`.
Now, for revolute joints, and frame B coincident with frame M, `v_PB_W = 0` and this is not even observable in the unit tests. 
Two options:
1. A test with prismatic joints (uff, we need those!)
2. A test where B != M, easier.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8194)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 7 of 11 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[multibody/multibody_tree/body_node.h, line 1235 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L75cE7SemZ_j-NtlTej:-L76sWufzRcmaO__rpsN:bmixd09) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1235)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Ok, I am pretty certain (I will push derivations soon to add here in this PR), that the Coriolis term here is off by a factor of 2. Regarding the rotational term, notice that: `w_WB x w_PB_W = w_WP x w_PB_W` and therefore that terms is nothing but the cross velocity term appearing when composing accelerations (more on this later when I update the notes).
Regarding the translational term notice that `v_WB = v_WP + w_WP x p_PoBo_W` and therefore the rotational component is: `w_WP x (v_WB - v_WP + v_PB_W) = w_WP x w_WP x p_PoBo_W + w_WP x v_PB_W`. The first part on the right hand side is the centrifugal acceleration while the second part should correspond to the Coriolis acceleration. However, Coriolis always comes with a factor of 2, i.e I believe this term should be `w_WP x w_WP x p_PoBo_W + 2 * w_WP x v_PB_W`.
Now, for revolute joints, and frame B coincident with frame M, `v_PB_W = 0` and this is not even observable in the unit tests. 
Two options:
1. A test with prismatic joints (uff, we need those!)
2. A test where B != M, easier.

</blockquote></details>

Never mind. Those two expressions are equivalent!! I made a mistake above. It should be: `v_WB = v_WP + w_WP x p_PoBo_W + v_PB_W`. Therefore we do get the factor of 2 we need for the Coriolis term. It is just strange to see it written in this way but, it is exactly the same!
As promised, I'll push some doc updates.
Also, I did try case (2) for the unit tests in `tree_from_joints.cc`, by just moving frame L2 around. And surprise, surprise I got the same results!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8194)*
<!-- Sent from Reviewable.io -->

bobbyluig commented:
Review status: 5 of 12 files reviewed at latest revision, 19 unresolved discussions.

---

*[multibody/multibody_tree/articulated_body_algorithm_cache.h, line 88 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6sQ3EYqVbev2BIUYbU:-L77pyEAQU_h4te1FSEr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/articulated_body_algorithm_cache.h#L88)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW blank line
</blockquote></details>

Done.

---

*[multibody/multibody_tree/articulated_body_algorithm_cache.h, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6sSHrdhBh6dF1USOzp:-L77q8Y-qT_mkyLKGgEy:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/articulated_body_algorithm_cache.h#L111)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW maybe place this comment in the line above just once instead of repeating on each line.
</blockquote></details>

Done.

---

*[multibody/multibody_tree/articulated_body_inertia_cache.h, line 123 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YB0BHw2LColPnkZNb:-L70Hc6T0llXP9c-DP1Z:bpr17mo) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/articulated_body_inertia_cache.h#L123)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

because of this initialization to NaN? where specifically? I could take a look at it.
</blockquote></details>

I'm currently trying `ldlt_D_B_[world_index()] = Matrix6<T>::Costant(nan()).ldlt()`.

---

*[multibody/multibody_tree/body_node.h, line 1048 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6sRTOg5ZNgRfzVE8_l:-L77qpD2neK6S08zCNoQ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1048)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW only --> on
</blockquote></details>

Done.

---

*[multibody/multibody_tree/body_node.h, line 1186 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6vgnD0FYB3Hn6hsxlQ:-L77vz5r8lwnySQRX6Gt:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1186)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW could you add to TODO comment this also gets used in `BodyNode::CalcBodySpatialForceGivenItsSpatialAcceleration()`?
</blockquote></details>

Done.

---

*[multibody/multibody_tree/body_node.h, line 1187 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6vgL41nRgdpSkzBqkA:-L77w05RamltpN7vI8Zv:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1187)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW `const`
</blockquote></details>

Done.

---

*[multibody/multibody_tree/multibody_tree.cc, line 755 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6qmQuWFyXnH92FchTg:-L77yECDJezP6XbuJDzP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/multibody_tree.cc#L755)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW blank line
</blockquote></details>

Done.

---

*[multibody/multibody_tree/multibody_tree.cc, line 969 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6qmyMcyyI3xIU1pRBB:-L785lFF6Sq_URofsXfD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/multibody_tree.cc#L969)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Actually, this is the perfect place for the acceleration cache (I initially had messed up how to use it).
So, on a second thought, why not to place `vdot` within the acceleration kinematics cache, and make the signature of the forward dynamics `CalcForwardDynamics(const Context&, const MultibodyForces&, AccelerationKinematicsCache*)`? That also gives us a chance to not only save `vdot` by also `A_WB`
</blockquote></details>

Done.

---

*[multibody/multibody_tree/multibody_tree.h, line 1445 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6qj7ojGYH_dh9-Zsvc:-L78-frA0N5geSCqF6A3:bzi1hnh) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/multibody_tree.h#L1445)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, not cited in the text. Did you mean to?
</blockquote></details>

Nope, removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8194)*
<!-- Sent from Reviewable.io -->

bobbyluig commented:
Review status: 3 of 14 files reviewed at latest revision, 10 unresolved discussions.

---

*[multibody/multibody_tree/articulated_body_inertia_cache.h, line 123 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6YB0BHw2LColPnkZNb:-L7q9bTe0XU-5EV5lS2g:b7aiv96) ([raw file](https://github.com/robotlocomotion/drake/blob/d0fb7cfbbe48ac05ffca869078e555538cc9ebb3/multibody/multibody_tree/articulated_body_inertia_cache.h#L123)):*
<details><summary><i>Previously, bobbyluig (Lujing Cen) wrote…</i></summary><blockquote>

I'm currently trying `ldlt_D_B_[world_index()] = Matrix6<T>::Costant(nan()).ldlt()`.
</blockquote></details>

See comment. It seems that LDLT does a comparison of abs(formula) > 0, which isn't supported by symbolic. 

---

*[multibody/multibody_tree/body_node.h, line 1115 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6sXkFbg7owKlSmxETb:-L7qBDO7F9BDSNfSsbvw:bg4bbg8) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1115)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

It doesn't seem like we need an extra symbol here. Recall that Jain uses `+` to denote the frame immediately inboard of the current body frame B. I think that frame for us is `Pb`, that is, body frame P shifted to be at Bo. Therefore I think instead of zplus (which is a spatial force) we could just use z_PPb_W (spatial force on P, at Po, expressed in W). That is, there's nothing special about zplus, it's just another bias term but at another frame (I think?). Look at Eq. 6.35 in page 108 of Jain's book, which says `z(k+1) = phi(k+1,k)*zplus(k)`. With this notation we'd simply get `Z_PPo_W = Z_PPb_W.Shift(p_BoPo_W)`.
Also, and this is just preference, maybe since z is a spatial force we could use upper case Z as we do for the upper case symbol F, i.e. `Z_PPb_W`. 
</blockquote></details>

Jain writes that z(k+1) = phi(k+1)*zplus(k), which I don't believe is the shift operator. For a parent body with a single child body, the ABA algorithm computes z(k) = phi(k, k-1)*zplus(k-1) + P(k)*a(k) + b(k) - external forces. I upper cased all forces. 

---

*[multibody/multibody_tree/body_node.h, line 1133 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6vd42lDd9qJWJW4kYn:-L7q7X1otYfCCZ2-6V8R:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1133)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

The derivations for these equations are quite lengthy but, could you at least cite an Eq. number in Jain's book?
</blockquote></details>

Done.

---

*[multibody/multibody_tree/body_node.h, line 1219 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6vhhPANckfrrytuS13:-L7qM9uTCobRgwr_2eWn:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1219)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This is not immediately obvious to someone who has not done this before. Could you add brief comment on how/why using `w_FM` instead of `w_PB`? (are they equal? why? assumptions).
Also comment on how the F frame in `Azero_FM` becomes the P frame in `Azero_PB_W` (that is, why is it that `A_PM = A_FM`?) (I actually had seconds thoughts myself for a moment).
</blockquote></details>

Done.

---

*[multibody/multibody_tree/body_node.h, line 1235 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L75cE7SemZ_j-NtlTej:-L7qE71BnU2IVttQGnTX:bs65ufi) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1235)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Never mind. Those two expressions are equivalent!! I made a mistake above. It should be: `v_WB = v_WP + w_WP x p_PoBo_W + v_PB_W`. Therefore we do get the factor of 2 we need for the Coriolis term. It is just strange to see it written in this way but, it is exactly the same!
As promised, I'll push some doc updates.
Also, I did try case (2) for the unit tests in `tree_from_joints.cc`, by just moving frame L2 around. And surprise, surprise I got the same results!
</blockquote></details>

We definitely need another test case. I will look into cleaning up this part.

---

*[multibody/multibody_tree/body_node.h, line 1332 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6s-tYPDqiJMmqH3G8J:-L7qFOtcl5_WIYILL5et:b-kh00jf) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1332)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Jain calls this Aplus because he counts from 1 to n from tip to base. This acceleration however is the acceleration of frame Pb (frame P shifted to b) if the system was at rest. Maybe we could call it something like `A0_WPb`?, that is, the acceleration of frame P shifted to B when the system is at rest (and thus velocities are zero).
</blockquote></details>

Agreed. The variable`Azero`may need to be changed, since it represents zero acceleration but may be confusing.

---

*[multibody/multibody_tree/body_node.h, line 1345 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6qpZYW70cp6ECLXsNn:-L7qEJNJltopTvyvulJN:b-cikyiz) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1345)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, is a "zero acceleration" a pre-requisite? I mean, could this entire algorithm work in a non-inertial frame of reference where the acceleration of the world is non-zero? For instance, imagine your robot is operating inside a moving vehicle (or the robot IS the vehicle). 
</blockquote></details>

No, that's not a prerequisite. Removed.

---

*[multibody/multibody_tree/body_node.h, line 1351 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6qrlfPBcZTLiup_c6X:-L7qEZuPr5Wj69B2qxKE:brdkrah) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/body_node.h#L1351)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

is there a reason to think about gravity differently from other forces? maybe say: "Note that we assume that any external forces have already been incorporated in the previous pass of the articulated body algorithm."
</blockquote></details>

Nope. It's just that Jain wrote it as a separate term. Updated docs.

---

*[multibody/multibody_tree/multibody_tree.h, line 1891 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6sQbSB9eXrz7qspEG8:-L7qODJFDpaADUJksYrh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/multibody_tree.h#L1891)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This and the doc for `CalcArticulatedBodyAlgorithmCache()` are very similar. I think it is worth mentioning in these docs that `CalcArticulatedBodyInertiaCache()` only depends on q and therefore solving forward dynamics with different forces, but the same context,  would require multiple evaluations of `CalcArticulatedBodyAlgorithmCache()` but only a single call to `CalcArticulatedBodyInertiaCache()`
</blockquote></details>

Done.

---

*[multibody/multibody_tree/test/tree_from_joints_test.cc, line 381 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8194#-L6qkDJ1_vEvA9lxgupF:-L7qF42V9IvX4VVO1xI6:b-p4ybu) ([raw file](https://github.com/robotlocomotion/drake/blob/64130c1def63d778b782ad8ef552a10fec829716/multibody/multibody_tree/test/tree_from_joints_test.cc#L381)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW comment is not correct?
</blockquote></details>

Oops, fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8194)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@amcastro-tri This is review blocked on you?  What's the story?

amcastro-tri commented:
Absolutely blocked on me. I upped it's priority (moved to "Ready for work") so that I can work on it asap.

jwnimmer-tri commented:
@amcastro-tri re-ping.

jwnimmer-tri commented:
@amcastro-tri re-ping.

amcastro-tri commented:
I've been working on this though we won't merge this PR as is since we need FD in an operator form to make use of effective caching and incorporate it in our contact solvers, see #11389.
